rust/lance-encoding/src/repdef.rs

```diff
@@ -541,6 +541,442 @@ impl RepDefUnraveler {
     }
 }
 
+/// A [`ControlWordIterator`] when there are both repetition and definition levels
+///
+/// The iterator will put the repetition level in the upper bits and the definition
+/// level in the lower bits.  The number of bits used for each level is determined
+/// by the width of the repetition and definition levels.
+#[derive(Debug)]
+pub struct BinaryControlWordIterator<I: Iterator<Item = (u16, u16)>, W> {
+    repdef: I,
+    def_width: usize,
+    rep_mask: u16,
+    def_mask: u16,
+    bits_rep: u8,
+    bits_def: u8,
+    phantom: std::marker::PhantomData<W>,
+}
+
+impl<I: Iterator<Item = (u16, u16)>> BinaryControlWordIterator<I, u8> {
+    fn append_next(&mut self, buf: &mut Vec<u8>) {
+        let next = self.repdef.next().unwrap();
+        let control_word: u8 =
+            (((next.0 & self.rep_mask) as u8) << self.def_width) + ((next.1 & self.def_mask) as u8);
+        buf.push(control_word);
+    }
+}
+
+impl<I: Iterator<Item = (u16, u16)>> BinaryControlWordIterator<I, u16> {
+    fn append_next(&mut self, buf: &mut Vec<u8>) {
+        let next = self.repdef.next().unwrap();
+        let control_word: u16 =
+            ((next.0 & self.rep_mask) << self.def_width) + (next.1 & self.def_mask);
+        let control_word = control_word.to_le_bytes();
+        buf.push(control_word[0]);
+        buf.push(control_word[1]);
+    }
+}
+
+impl<I: Iterator<Item = (u16, u16)>> BinaryControlWordIterator<I, u32> {
+    fn append_next(&mut self, buf: &mut Vec<u8>) {
+        let next = self.repdef.next().unwrap();
+        let control_word: u32 = (((next.0 & self.rep_mask) as u32) << self.def_width)
+            + ((next.1 & self.def_mask) as u32);
+        let control_word = control_word.to_le_bytes();
+        buf.push(control_word[0]);
```

Comment:
is there any reason we don't like to use extend_from_slice?
```rust
buf.extend_from_slice(control_word.to_le_bytes())
```

Reply:
Maybe I'm just being paranoid.  However, my reasoning was that `buf.extend_from_slice` will invoke a `memcpy` where `buf.push` should not.  However, there may be a fast path in `extend_from_slice` somewhere or it may get optimized away or...